walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,180 @@
+  ) {
+    // Potential windows that only include mandatory attendees.
+    ArrayList<TimeRange> timeSlots = new ArrayList<>();
+    // Potential windows that include ALL mandatory + optional attendees.
+    ArrayList<TimeRange> timeSlotsWithOptional = new ArrayList<>();
+
+    ArrayList<Event> sortedEvents = new ArrayList<>(events);
+    sortedEvents.sort(EVENT_COMPARATOR);
+
+    /*
+     * We consider this problem in terms of 'windows' and 'blocks'.
+     * |-----A-----|                  |------A-----|
+     *       |----------B-----|       |---B---|
+     *
+     * |______________________||______|
+     *        block              window
+     *
+     * We iterate shifts to identify the right-most extent of a block.
+     * If we find a shift that start *after* this point, the difference is a window.
+     * That window is a possible meeting spot, if long enough.
+     */
+    int mandatoryBlockExtent = TimeRange.START_OF_DAY; // Based on people who *must* attend
+    int optionalBlockExtent = TimeRange.START_OF_DAY; // Based on people who *may* attend
+
+    // Whether any events exist for mandatory and optional attendees respectively.
+    boolean hasMandatoryEvents = false;
+    boolean hasOptionalEvents = false;
+
+    // Whether we should consider attendees of a given class.
+    // This is used to handle cases like all optional no mandatory, no attendees, etc.
+    boolean considerMandatoryAttendees = request.getAttendees().size() > 0;
+    boolean considerOptionalAttendees = request.getOptionalAttendees().size() > 0;
+
+    for (Event event : sortedEvents) {
+      TimeRange when = event.getWhen();
+      // If true, the meeting request cannot overlap with this.
+      boolean containsMandatoryAttendees = anyAttendeesAreMandatory(
+        event,
+        request
+      );
+      // If true, meeting requests that try to include optional attendees cannot overlap with this.
+      boolean containsOptionalAttendees = anyAttendeesAreOptional(
+        event,
+        request
+      );
+
+      if (containsMandatoryAttendees) {
+        hasMandatoryEvents = true;
+      }
+      if (containsOptionalAttendees) {
+        hasOptionalEvents = true;
+      }
+
+      // As mentioned above, we track 'mandatory' people and 'optional' people as two classes.
+      // For a given class, a window exists if there is time between current event involving that class, 
+      // And the end of the last event involving that class (_____BlockExtent)
+      boolean optionalWindowExists = when.start() > optionalBlockExtent;
+      boolean mandatoryWindowExists = when.start() > mandatoryBlockExtent;
+
+      // Makes sure the requested meeting length fits within the window.
+      boolean optionalWindowLongEnough = optionalWindowExists && windowIsLongEnough(optionalBlockExtent, when.start(), request);
+      boolean mandatoryWindowLongEnough = mandatoryWindowExists && windowIsLongEnough(mandatoryBlockExtent, when.start(), request);
+
```

Comment:
While this isn't the typical duplicated code, you're writing the same code over and over for the optional and mandatory cases.

No need to change your code, but the thought that comes to me is that instead of doing twice the work in one iteration through the events, you could do two iterations: once for all attendees, and once for mandatory attendees if the first iteration failed to yield results.